File: sdk/typescript/examples/06-matrix/sykli.ts
/**
 * Example 06: Matrix Builds and Advanced Features
 *
 * This example demonstrates:
 * - Matrix builds across multiple dimensions
 * - Dynamic pipeline generation
 * - Kubernetes options
 * - Secrets management
 * - Retry and timeout
 *
 * Run with: npx ts-node --esm sykli.ts --emit | sykli run -
 */

import { Pipeline, branch, tag, fromEnv, fromVault } from 'sykli';

const p = new Pipeline();

// === MATRIX BUILDS ===
// Test across Node versions
const nodeVersions = ['18', '20', '22'];

const nodeTests = p.matrix('node-tests', nodeVersions, (version) =>
  p.task(`test-node-${version}`)
    .container(`node:${version}`)
    .mountCwd()
    .run('npm ci && npm test')
);

// Matrix with map: different browsers
const browsers: Record<string, string> = {
  chrome: 'mcr.microsoft.com/playwright:latest',
  firefox: 'mcr.microsoft.com/playwright:latest',
  webkit: 'mcr.microsoft.com/playwright:latest',
};

const e2eTests = p.matrixMap('e2e-tests', browsers, (browser, image) =>
  p.task(`e2e-${browser}`)
    .container(image)
    .mountCwd()
    .env('BROWSER', browser)
    .run('npx playwright test')
    .afterGroup(nodeTests)
);

// === DYNAMIC GENERATION ===
// Generate tasks from data (TypeScript power!)
const services = ['api', 'web', 'worker'];

for (const svc of services) {
  p.task(`build-${svc}`)
    .container('node:20')
    .mountCwd()
    .run(`npm run build:${svc}`)
    .afterGroup(nodeTests);
}

// === KUBERNETES OPTIONS ===
p.task('gpu-inference')
  .container('pytorch/pytorch:2.0')
  .mountCwd()
  .run('python inference.py')
  .k8s({
    gpu: 1,
    resources: {
      requestMemory: '8Gi',
      limitMemory: '16Gi',
    },
    nodeSelector: { 'gpu-type': 'nvidia-a100' },
    tolerations: [
      { key: 'nvidia.com/gpu', operator: 'Exists', effect: 'NoSchedule' },
    ],
  })
  .afterGroup(nodeTests);

// === SECRETS ===
p.task('deploy')
  .container('alpine:3.18')
  .run('./deploy.sh')
  .secret('DEPLOY_TOKEN') // Simple secret
  .secretFrom('AWS_KEY', fromEnv('AWS_ACCESS_KEY_ID'))
  .secretFrom('DB_PASS', fromVault('secret/data/prod/db'))
  .whenCond(branch('main'))
  .afterGroup(e2eTests);

// === RETRY & TIMEOUT ===
p.task('flaky-integration')
  .container('node:20')
  .mountCwd()
  .run('npm run test:integration')
  .retry(3) // Retry up to 3 times
  .timeout(600) // 10 minute timeout
  .afterGroup(nodeTests);

// === PUBLISH ===
p.task('publish')
  .run('npm publish')
  .secret('NPM_TOKEN')
  .whenCond(tag('v*'))
  .afterGroup(e2eTests);

p.emit();

// Matrix expansion:
//
// test-node-18 ─┐
// test-node-20 ─┼─▶ e2e-chrome  ─┐
// test-node-22 ─┘   e2e-firefox ─┼─▶ deploy (on main)
//                   e2e-webkit  ─┘   publish (on v*)

File: sdk/go/examples/06-matrix/sykli.go
//go:build ignore

// Example 06: Matrix Builds & Services
//
// This example demonstrates:
// - Matrix() for testing across configurations
// - MatrixMap() for named matrix variations
// - Service() for background containers
// - Retry() and Timeout() for resilience
//
// Run with: sykli run

package main

import sykli "github.com/yairfalse/sykli/sdk/go"

func main() {
	s := sykli.New()

	src := s.Dir(".")

	// === MATRIX BUILDS ===

	// Test across multiple Go versions
	// Creates: test-1.20, test-1.21, test-1.22
	goVersions := s.Matrix("go-test", []string{"1.20", "1.21", "1.22"}, func(version string) *sykli.Task {
		return s.Task("test-go-"+version).
			Container("golang:"+version).
			Mount(src, "/src").
			MountCache(s.Cache("go-mod-"+version), "/go/pkg/mod").
			Workdir("/src").
			Run("go test ./...")
	})

	// === NAMED MATRIX (MatrixMap) ===

	// Deploy to multiple environments
	// Creates: deploy-staging, deploy-prod
	s.MatrixMap("deploy", map[string]string{
		"staging": "staging.example.com",
		"prod":    "prod.example.com",
	}, func(env, host string) *sykli.Task {
		return s.Task("deploy-"+env).
			Run("./deploy.sh --host " + host).
			When("branch == 'main'").
			AfterGroup(goVersions)
	})

	// === SERVICE CONTAINERS ===

	// Integration tests with database and cache services
	s.Task("integration").
		Container("golang:1.21").
		Mount(src, "/src").
		Workdir("/src").
		Service("postgres:15", "db").      // Accessible as hostname "db"
		Service("redis:7", "cache").       // Accessible as hostname "cache"
		Env("DATABASE_URL", "postgres://postgres:postgres@db:5432/test?sslmode=disable").
		Env("REDIS_URL", "redis://cache:6379").
		Run("go test -tags=integration ./...").
		Timeout(300).  // 5 minute timeout
		AfterGroup(goVersions)

	// === RETRY & TIMEOUT ===

	// Flaky tests with retry
	s.Task("e2e").
		Container("cypress/included:13").
		Mount(src, "/src").
		Workdir("/src").
		Run("npm run e2e").
		Retry(3).      // Retry up to 3 times
		Timeout(600).  // 10 minute timeout
		After("integration")

	// === SECRETS ===

	// Deployment with secrets
	s.Task("publish").
		Run("./publish.sh").
		Secret("NPM_TOKEN").
		SecretFrom("GITHUB_TOKEN", sykli.FromEnv("GH_TOKEN")).
		When("tag != ''").
		After("e2e")

	s.Emit()
}

// Generated tasks:
//
// test-go-1.20  ─┐
// test-go-1.21  ─┼─> integration ─> e2e ─> publish
// test-go-1.22  ─┤
//                └─> deploy-staging
//                └─> deploy-prod
